code/game/objects/items/storage/boxes.dm

```diff
@@ -536,3 +608,69 @@
 	if(H == usr && !H.incapacitated() && Adjacent(H) && H.put_in_hands(src))
 		deployed = FALSE
 		update_icon()
+
+/obj/item/storage/box/magazine/update_icon_state()
+	. = ..()
+
+	//Clear all overlays for a fresh start
+	for(var/I in overlays)
+		overlays -= I
+	variety = 0
+
+	//Create an assoc list of every ammo type in the crate and have it's value be the total weight it takes up.
+	contents_weight = list()
+	for(var/obj/item/I in contents)
+		if(!contents_weight[I.type])
+			contents_weight[I.type] = 0
+			variety += 1
+		contents_weight[I.type] += I.w_class
+
+	if(!deployed)
+		icon_state = "[initial(icon_state)]"
+		return
+	if(variety > max_overlays) // Too many items inside so lets make it cluttered
+		icon_state = "[initial(icon_state)]_mixed"
+		return
+
+	icon_state = "[initial(icon_state)]_open"
+
+	//Warning: Shitcode ahead.
+
+	//Determine the amount of overlays to draw
+	var/total_overlays = 0
+	for(var/W in contents_weight)
+		total_overlays += 1 + FLOOR(contents_weight[W] / overlay_w_class, 1)
+
+	//In case 6 overlays are for a LMG and then someone adds 7 unique tiny items into the mix
+	var/overlay_overflow = max(0, total_overlays - max_overlays)
+
+	//The Xth overlay being drawed.
+	var/current_iteration = 1
+
+	//Fun.exe has determined that the following lines contain 100% hardcoded shitcode. Maintainer discretion is advised.
+	for(var/W in contents_weight) //Max 8 items in contents_weight since otherwise the icon_state would be "mixed"
+		var/overlays_to_draw = 1 + FLOOR(contents_weight[W] / overlay_w_class, 1) //Always draw at least 1 icon per unique mag and add additional icons if it takes a lot of space inside.
+		if(overlay_overflow)//This makes sure no matter the configuration, every item will get at least 1 spot in the mix.
+			var/adjustment = min(overlay_overflow, overlays_to_draw - 1)
+			overlay_overflow -= adjustment
+			overlays_to_draw -= adjustment
+			total_overlays -= adjustment
+		for(var/i = 1, i <= overlays_to_draw, i++)
+			var/imagepixel_x = overlay_pixel_x + FLOOR((current_iteration / 2) - 0.5, 1) * shift_x
+			var/imagepixel_y = overlay_pixel_y
+			if(MODULUS(current_iteration, 2) && (total_overlays - current_iteration) > 0) // We draw the top part first to account for overlaps.
+				imagepixel_y += shift_y
+
+			//Getting the mini icon_state to display
+			var/obj/item/ammo_magazine/relatedmag = W
+			var/imagestate =  initial(relatedmag?.icon_state_mini)
+			if(!imagestate) //Might be a cell or a random item
+				for(var/A in assoc_overlay)
+					message_admins("[W] - [A]")
+					if(findtext("[W]", "[A]")) //istype is wacky
+						imagestate = assoc_overlay[A]
+						break
```

Comment:
was all this really the best yeah to achieve this result?

Reply:
When i have two assoc lists with the keys being the types and i try to do istype(key, key) on them, they always fail.
Might be byond might be me being a dung.
Who knows.
Gonna look more into it.